mars/learn/contrib/tensorflow/dataset.py

```diff
@@ -0,0 +1,160 @@
+# Copyright 1999-2020 Alibaba Group Holding Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import copy
+from typing import List, Tuple
+
+import pandas as pd
+import numpy as np
+
+from .... import execute
+from ....core.context import get_context
+from ....tensor.core import TENSOR_TYPE
+from ....dataframe.core import DATAFRAME_TYPE, SERIES_TYPE
+from ....utils import require_not_none, lazy_import
+
+tf = lazy_import("tensorflow")
+
+
+ACCEPT_TYPE = (TENSOR_TYPE, DATAFRAME_TYPE, SERIES_TYPE,
+               np.ndarray, pd.DataFrame, pd.Series, List)
+
+
+@require_not_none(tf)
+class MarsDataset:
+    def __init__(self, tensors,
+                 output_shapes = None,
+                 output_types = None,
+                 fetch_kwargs=None):
+
+        self._context = get_context()
+        self._tensors = tensors
+        self._output_shapes = output_shapes
+        self._output_types = output_types
+        self._fetch_kwargs = fetch_kwargs or dict()
+        self._executed = False
+        self._check_and_convert()
+
+    def _check_and_convert(self):
+        if not isinstance(self._tensors, Tuple):
+            self._tensors = (self._tensors,)
+        for t in self._tensors:
+            if not isinstance(t, ACCEPT_TYPE):
+                raise TypeError(f"Unexpected dataset type: {type(t)}")
+
+        if not self._output_shapes:
+            get_shape = lambda t: tuple(()) if isinstance(t, (List, SERIES_TYPE, pd.Series)) \
+                                  else t.shape[1:]
+            self._output_shapes = get_shape(self._tensors[0]) if len(self._tensors) == 1 else \
```

Comment:
Better to get shape after `.execute()`, because sometimes, some tensor or DataFrame has unknown shape whose shape will be like `(np,nan,)`, the real shape will be updated after `.execute()`.